File: backend/services/resource_permission_service.py
from typing import List, Dict, Any, Optional
from datetime import datetime
from bson import ObjectId
from motor.motor_asyncio import AsyncIOMotorDatabase

from models.foursyz.permissions import (
    ResourcePermission, UserResourceAssignment, PermissionEvaluation,
    PREDEFINED_RESOURCE_PERMISSIONS, RESOURCE_PERMISSION_CATEGORIES,
    Action, Resource, Effect
)

class ResourcePermissionService:
    def __init__(self, db: AsyncIOMotorDatabase):
        self.db = db
        self.resource_permissions_collection = db.resource_permissions
        self.user_resource_assignments_collection = db.user_resource_assignments

    async def initialize_resource_permissions(self) -> Dict[str, Any]:
        """Initialize predefined resource permissions in the database"""
        try:
            # Check if permissions already exist
            existing_count = await self.resource_permissions_collection.count_documents({})
            if existing_count > 0:
                return {
                    "status": "info",
                    "message": f"Resource permissions already initialized ({existing_count} permissions exist)",
                    "permissions_created": 0
                }

            # Insert predefined permissions
            permissions_to_insert = []
            for permission_id, permission in PREDEFINED_RESOURCE_PERMISSIONS.items():
                permission_dict = permission.dict()
                permission_dict["created_at"] = datetime.utcnow()
                permission_dict["updated_at"] = datetime.utcnow()
                permissions_to_insert.append(permission_dict)

            if permissions_to_insert:
                result = await self.resource_permissions_collection.insert_many(permissions_to_insert)
                return {
                    "status": "success",
                    "message": f"Successfully initialized {len(result.inserted_ids)} resource permissions",
                    "permissions_created": len(result.inserted_ids)
                }
            else:
                return {
                    "status": "warning",
                    "message": "No predefined permissions to initialize",
                    "permissions_created": 0
                }

        except Exception as e:
            return {
                "status": "error",
                "message": f"Failed to initialize resource permissions: {str(e)}",
                "permissions_created": 0
            }

    async def get_all_resource_permissions(self) -> List[ResourcePermission]:
        """Get all resource permissions"""
        try:
            cursor = self.resource_permissions_collection.find({})
            permissions = []
            async for doc in cursor:
                permissions.append(ResourcePermission(**doc))
            return permissions
        except Exception as e:
            raise Exception(f"Failed to get resource permissions: {str(e)}")

    async def get_resource_permissions_by_category(self) -> Dict[str, List[ResourcePermission]]:
        """Get resource permissions grouped by category"""
        try:
            permissions = await self.get_all_resource_permissions()
            categorized = {}
            
            for permission in permissions:
                if permission.category not in categorized:
                    categorized[permission.category] = []
                categorized[permission.category].append(permission)
            
            return categorized
        except Exception as e:
            raise Exception(f"Failed to get categorized permissions: {str(e)}")

    async def get_resource_permission_by_id(self, permission_id: str) -> Optional[ResourcePermission]:
        """Get a specific resource permission by ID"""
        try:
            doc = await self.resource_permissions_collection.find_one({"id": permission_id})
            if doc:
                return ResourcePermission(**doc)
            return None
        except Exception as e:
            raise Exception(f"Failed to get resource permission: {str(e)}")

    async def assign_resource_permission_to_user(
        self, 
        user_id: str, 
        permission_id: str, 
        assigned_by: str,
        expires_at: Optional[datetime] = None,
        notes: Optional[str] = None
    ) -> UserResourceAssignment:
        """Assign a resource permission to a user"""
        try:
            # Check if permission exists
            permission = await self.get_resource_permission_by_id(permission_id)
            if not permission:
                raise Exception(f"Resource permission '{permission_id}' not found")

            # Check if assignment already exists and is active
            existing_assignment = await self.user_resource_assignments_collection.find_one({
                "user_id": user_id,
                "resource_permission_id": permission_id,
                "active": True
            })

            if existing_assignment:
                raise Exception(f"User already has active assignment for permission '{permission_id}'")

            # Create new assignment
            assignment = UserResourceAssignment(
                user_id=user_id,
                resource_permission_id=permission_id,
                assigned_at=datetime.utcnow(),
                assigned_by=assigned_by,
                expires_at=expires_at,
                notes=notes,
                active=True,
                created_at=datetime.utcnow(),
                updated_at=datetime.utcnow()
            )

            assignment_dict = assignment.dict()
            result = await self.user_resource_assignments_collection.insert_one(assignment_dict)
            assignment.id = str(result.inserted_id)

            return assignment

        except Exception as e:
            raise Exception(f"Failed to assign resource permission: {str(e)}")

    async def unassign_resource_permission_from_user(
        self, 
        user_id: str, 
        permission_id: str
    ) -> bool:
        """Unassign a resource permission from a user"""
        try:
            result = await self.user_resource_assignments_collection.update_many(
                {
                    "user_id": user_id,
                    "resource_permission_id": permission_id,
                    "active": True
                },
                {
                    "$set": {
                        "active": False,
                        "updated_at": datetime.utcnow()
                    }
                }
            )

            if result.modified_count > 0:
                return True
            else:
                raise Exception(f"No active assignment found for user '{user_id}' and permission '{permission_id}'")

        except Exception as e:
            raise Exception(f"Failed to unassign resource permission: {str(e)}")

    async def get_user_resource_assignments(self, user_id: str) -> List[UserResourceAssignment]:
        """Get all resource permission assignments for a user"""
        try:
            cursor = self.user_resource_assignments_collection.find({
                "user_id": user_id,
                "active": True
            })
            
            assignments = []
            async for doc in cursor:
                assignments.append(UserResourceAssignment(**doc))
            
            return assignments
        except Exception as e:
            raise Exception(f"Failed to get user resource assignments: {str(e)}")

    async def get_user_resource_permissions(self, user_id: str) -> List[ResourcePermission]:
        """Get all resource permissions assigned to a user"""
        try:
            assignments = await self.get_user_resource_assignments(user_id)
            permissions = []
            
            for assignment in assignments:
                permission = await self.get_resource_permission_by_id(assignment.resource_permission_id)
                if permission:
                    permissions.append(permission)
            
            return permissions
        except Exception as e:
            raise Exception(f"Failed to get user resource permissions: {str(e)}")

    async def evaluate_user_permission(
        self, 
        user_id: str, 
        action: Action, 
        resource: str,
        context: Optional[Dict[str, Any]] = None
    ) -> PermissionEvaluation:
        """Evaluate if a user has permission to perform an action on a resource"""
        try:
            # Get user's resource permissions
            user_permissions = await self.get_user_resource_permissions(user_id)
            
            if not user_permissions:
                return PermissionEvaluation(
                    allowed=False,
                    reason="User has no resource permissions assigned",
                    required_action=action,
                    required_resource=resource,
                    evaluated_policies=[]
                )

            # Check each permission for matching action and resource
            for permission in user_permissions:
                # Check if action is allowed
                if action in permission.actions:
                    # Check if resource matches (support wildcards)
                    if self._resource_matches(permission.resource, resource):
                        return PermissionEvaluation(
                            allowed=True,
                            reason=f"Permission granted by '{permission.id}'",
                            matched_statement=None,  # Not applicable for resource permissions
                            evaluated_policies=[permission.id],
                            required_action=action,
                            required_resource=resource
                        )

            return PermissionEvaluation(
                allowed=False,
                reason=f"Action '{action}' not allowed on resource '{resource}'",
                required_action=action,
                required_resource=resource,
                evaluated_policies=[p.id for p in user_permissions]
            )

        except Exception as e:
            return PermissionEvaluation(
                allowed=False,
                reason=f"Error evaluating permission: {str(e)}",
                required_action=action,
                required_resource=resource,
                evaluated_policies=[]
            )

    def _resource_matches(self, permission_resource: str, requested_resource: str) -> bool:
        """Check if a requested resource matches a permission resource pattern"""
        # Exact match
        if permission_resource == requested_resource:
            return True
        
        # Wildcard match (e.g., "user:*" matches "user:123")
        if permission_resource.endswith(":*"):
            base_resource = permission_resource[:-2]  # Remove ":*"
            if requested_resource.startswith(base_resource + ":"):
                return True
        
        return False

    async def get_all_user_assignments(self) -> List[UserResourceAssignment]:
        """Get all active user resource assignments"""
        try:
            cursor = self.user_resource_assignments_collection.find({"active": True})
            assignments = []
            async for doc in cursor:
                assignments.append(UserResourceAssignment(**doc))
            return assignments
        except Exception as e:
            raise Exception(f"Failed to get all user assignments: {str(e)}")

    async def migrate_from_policy_system(self) -> Dict[str, Any]:
        """Migrate existing policy assignments to resource-based permissions"""
        try:
            # This would be implemented to migrate existing policy assignments
            # For now, return a placeholder
            return {
                "status": "info",
                "message": "Migration from policy system not yet implemented",
                "migrated_count": 0
            }
        except Exception as e:
            return {
                "status": "error",
                "message": f"Migration failed: {str(e)}",
                "migrated_count": 0
            } 

File: frontend/src/utils/resourcePermissionService.ts
import { env } from '@/config/env'

// Types for resource permissions
export interface ResourcePermission {
  id: string
  resource: string
  actions: string[]
  description: string
  category: string
  created_at?: string
  updated_at?: string
}

export interface UserResourceAssignment {
  id?: string
  user_id: string
  resource_permission_id: string
  assigned_at: string
  assigned_by: string
  expires_at?: string
  notes?: string
  active: boolean
  created_at?: string
  updated_at?: string
}

export interface PermissionEvaluation {
  allowed: boolean
  reason: string
  matched_statement?: Record<string, unknown>
  evaluated_policies: string[]
  required_action?: string
  required_resource?: string
}

// Request/Response types
export interface ResourcePermissionListResponse {
  status: string
  message: string
  permissions: ResourcePermission[]
  total: number
}

export interface UserResourceAssignmentListResponse {
  status: string
  message: string
  assignments: UserResourceAssignment[]
  total: number
}

export interface UserResourcePermissionsResponse {
  status: string
  message: string
  user_id: string
  permissions: ResourcePermission[]
  assignments: UserResourceAssignment[]
}

export interface PermissionEvaluationRequest {
  user_id: string
  action: string
  resource: string
  context?: Record<string, unknown>
}

export interface PermissionEvaluationResponse {
  status: string
  message: string
  evaluation: PermissionEvaluation
}

export interface UserResourceAssignmentRequest {
  assigned_by: string
  expires_at?: string
  notes?: string
}

export interface AssignmentData {
  assigned_by: string
  expires_at?: string
  notes?: string
}

export interface PermissionEvaluationData {
  user_id: string
  action: string
  resource: string
  context?: Record<string, unknown>
}

class ResourcePermissionService {
  private baseUrl: string

  constructor() {
    this.baseUrl = env.API_BASE_URL
  }

  private async makeRequest<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const token = localStorage.getItem(env.STORAGE_KEYS.ACCESS_TOKEN)
    
    const config: RequestInit = {
      headers: {
        'Content-Type': 'application/json',
        ...(token && { Authorization: `Bearer ${token}` }),
        ...options.headers,
      },
      ...options,
    }

    const response = await fetch(`${this.baseUrl}${endpoint}`, config)

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}))
      throw new Error(errorData.detail || `HTTP error! status: ${response.status}`)
    }

    return response.json()
  }

  // Initialize resource permissions
  async initializeResourcePermissions(): Promise<{ status: string; message: string; permissions_created: number }> {
    return this.makeRequest('/resource-permissions/initialize', {
      method: 'POST',
    })
  }

  // Get all resource permissions
  async getResourcePermissions(): Promise<ResourcePermission[]> {
    const response = await this.makeRequest<ResourcePermissionListResponse>('/resource-permissions/')
    return response.permissions
  }

  // Get resource permissions by category
  async getResourcePermissionsByCategory(): Promise<Record<string, ResourcePermission[]>> {
    return this.makeRequest<Record<string, ResourcePermission[]>>('/resource-permissions/categorized')
  }

  // Get specific resource permission
  async getResourcePermission(permissionId: string): Promise<ResourcePermission> {
    return this.makeRequest<ResourcePermission>(`/resource-permissions/${permissionId}`)
  }

  // Assign resource permission to user
  async assignResourcePermissionToUser(
    userId: string,
    permissionId: string,
    assignmentData: AssignmentData
  ): Promise<{ status: string; message: string; assignment: UserResourceAssignment }> {
    return this.makeRequest(`/resource-permissions/assign/${userId}/${permissionId}`, {
      method: 'POST',
      body: JSON.stringify(assignmentData),
    })
  }

  // Unassign resource permission from user
  async unassignResourcePermissionFromUser(
    userId: string,
    permissionId: string
  ): Promise<{ status: string; message: string }> {
    return this.makeRequest(`/resource-permissions/unassign/${userId}/${permissionId}`, {
      method: 'DELETE',
    })
  }

  // Get user resource assignments
  async getUserResourceAssignments(userId: string): Promise<UserResourceAssignment[]> {
    const response = await this.makeRequest<UserResourceAssignmentListResponse>(
      `/resource-permissions/user/${userId}/assignments`
    )
    return response.assignments
  }

  // Get user resource permissions
  async getUserResourcePermissions(userId: string): Promise<{
    permissions: ResourcePermission[]
    assignments: UserResourceAssignment[]
  }> {
    const response = await this.makeRequest<UserResourcePermissionsResponse>(
      `/resource-permissions/user/${userId}/permissions`
    )
    return {
      permissions: response.permissions,
      assignments: response.assignments,
    }
  }

  // Evaluate permission
  async evaluatePermission(evaluationData: PermissionEvaluationData): Promise<PermissionEvaluation> {
    const response = await this.makeRequest<PermissionEvaluationResponse>(
      '/resource-permissions/evaluate',
      {
        method: 'POST',
        body: JSON.stringify(evaluationData),
      }
    )
    return response.evaluation
  }

  // Get all user assignments
  async getAllUserAssignments(): Promise<UserResourceAssignment[]> {
    const response = await this.makeRequest<UserResourceAssignmentListResponse>(
      '/resource-permissions/assignments/all'
    )
    return response.assignments
  }

  // Get users (assuming this is available from auth service)
  async getUsers(): Promise<Record<string, unknown>[]> {
    // This would typically come from a user service
    // For now, we'll need to implement this or use existing auth service
    throw new Error('getUsers method needs to be implemented or use existing auth service')
  }

  // Helper method to check if user has permission
  async hasPermission(userId: string, action: string, resource: string): Promise<boolean> {
    try {
      const evaluation = await this.evaluatePermission({
        user_id: userId,
        action,
        resource,
      })
      return evaluation.allowed
    } catch (error) {
      console.error('Error checking permission:', error)
      return false
    }
  }

  // Helper method to get user's permissions for navigation
  async getUserPermissionsForNavigation(userId: string): Promise<ResourcePermission[]> {
    try {
      const { permissions } = await this.getUserResourcePermissions(userId)
      return permissions
    } catch (error) {
      console.error('Error getting user permissions for navigation:', error)
      return []
    }
  }
}

// Export singleton instance
export const resourcePermissionService = new ResourcePermissionService() 